### ISSUE TITLE
Testing Ryu-Backend + Core + Ryu-shim

### ISSUE THREAD

**Issue by andres-beato** [fp7-netide/Engine#42]
Hi all,
I have tested the Ryu-Backend and the Ryu-shim with the AdvancedProxyCore.py and they seem to work well.

Then, I tried to test the Ryu backend and Shim with the complete Core and it doesn't work, I guess I need to add a composition file, is it so? Could you give an example to do the composition file?

**Comment by sergiotamu** [comment]
Hi Andrés,

I have also problems with the Core. I don't know how to install and deploy it.

**Comment by TimVinkemeier** [comment]
@andres-beato Yes, you need to provide a CompositionSpecification, otherwise the core will drop all events and log something like "Unable to handle request due to configuration error". You can find an example specification here https://github.com/fp7-netide/Engine/tree/CoreImplementation/core/specification. You have to provide the file using the emulator tool (https://github.com/fp7-netide/Engine/tree/CoreImplementation/core/tools/emulator) and the `composition` command. It will ask for a file name, make sure that the file is in the same folder as the emulator executable jar.

@sergiotamu Although it is not fully up-to-date (you should use Karaf 3.0.3), the general instructions from https://github.com/fp7-netide/Engine/blob/CoreImplementation/core/Readme.md still hold.

**Comment by sergiotamu** [comment]
My problem was that I have installed the java version 1.7 instead of having installed 1.8. @TimVinkemeier

**Comment by andres-beato** [comment]
Hi @TimVinkemeier,
Sergio and I have been testing the Ryu shim and backend with the Complete core. We have used the following composition file to run the simple_switch.py over Ryu backend:

![composition_file](https://cloud.githubusercontent.com/assets/11061465/10316770/4d8d12d6-6c61-11e5-87a4-f809d7e0ca19.PNG)

Is it correct? Because we have in the core the following error:

2015-10-06 19:17:28,363 | INFO  | pool-6-thread-25 | CompositionManager               | 67 - core.caos - 1.0.0.0 | CompositionManager received message from shim: OpenFlowMessage [Header=MessageHeader [Version=VERSION_1_1,Type=OPENFLOW,Length=8,ModuleId=0,TransactionId=0,DatapathId=4],Type=ECHO_REQUEST,OFMessage=OFEchoRequestVer10(xid=0, data=[])]
2015-10-06 19:17:28,363 | ERROR | pool-6-thread-25 | CompositionManager               | 67 - core.caos - 1.0.0.0 | Could not handle incoming message due to configuration error.

Thanks in advance

**Comment by TimVinkemeier** [comment]
Hi @andres-beato,

yes, the composition file itself is correct, but does not define a composition :laughing: 
There is one module defined, but the `Composition` section is empty. Have a look at the more complete example https://github.com/fp7-netide/Engine/blob/CoreImplementation/core/specification/CompositionSpecification.xml and add at least a single ModuleCall.

**Comment by sergiotamu** [comment]
Hi @TimVinkemeier ,

I get the same error:

2015-10-06 22:08:51,833 | ERROR | pool-6-thread-7  | CompositionManager               | 67 - core.caos - 1.0.0.0 | Could not handle incoming message due to configuration error.

But now I am using this CompositionSpecification:
![composition](https://cloud.githubusercontent.com/assets/9716567/10320885/f9882076-6c75-11e5-931b-7eefa8d165a1.png)

Maybe, is it wrong?

**Comment by sergiotamu** [comment]
Hello @TimVinkemeier 

Now, we have not error but we can't ping the hosts deployed with mininet. The procedure that we follow is the next:
1) We execute Karaf
2)We execute the emulator and provide the Composition file to the Core.
3) With the command "a" (announcement) we indicate that the "sw" module will be executed.
4) We deploy the topology with mininet
5) The Ryu Shim is executed and is connected to the Core and to the data plane.
6) The Ryu backend and "Simple_switch" application are also executed and the backend is connected to the Core.
7) We use the command "ping" in mininet Console. However the hosts cannot ping each other.

We think maybe there is a problem in the communication between any of the components.
Do you think this process makes sense?? What do you think about??.

Thank you in advance.

**Comment by doriguzzi** [comment]
Hi Sergio,

the name of the module running in Ryu is "SimpleSwitch" not "sw". This could be a reason.
I left the original name of the Ryu app, so that you can use any Ryu application you want, without changing either the code of the backend or the code of the app. You just need to specify the right name in the composition specification.

Hope this helps.

--rob

**Comment by doriguzzi** [comment]
Usually the name of the Ryu apps is defined in the **init** method. 

E.g.: SimpleSwitch for the OF1.0 learning switch, SimpleSwitch13 for the OF1.3 learning switch, etc.

**Comment by TimVinkemeier** [comment]
Hi Sergio,

this obviously can't work. With the announcement command, you create a mapping between the module name and the backend that announced it, in your case the emulator. When the composition starts, the core routes the request to "sw" to the emulator, since it was announced there. Therefore, the real backend never receives it (you should however be able to see the request in the emulator console, if it is still connected as a backend). The real backend has to do the announcement itself and I think Roberto said that he already added the support for that...

Best,
Tim

----- Ursprüngliche Nachricht -----
Von: "Sergio Tamurejo" notifications@github.com
Gesendet: ‎07.‎10.‎2015 14:21
An: "fp7-netide/Engine" Engine@noreply.github.com
Cc: "Tim Vinkemeier" tim@vinkemeier.de
Betreff: Re: [Engine] Testing Ryu-Backend + Core + Ryu-shim (#42)

Hello @TimVinkemeier 
Now, we have not error but we can't ping the hosts deployed with mininet. The procedure that we follow is the next:
1) We execute Karaf
2)We execute the emulator and provide the Composition file to the Core.
3) With the command "a" (announcement) we indicate that the "sw" module will be executed.
4) We deploy the topology with mininet
5) The Ryu Shim is executed and is connected to the Core and to the data plane.
6) The Ryu backend and "Simple_switch" application are also executed and the backend is connected to the Core.
7) We use the command "ping" in mininet Console. However the hosts cannot ping each other.
We think maybe there is a problem in the communication between any of the components.
Do you think this process makes sense?? What do you think about.
Thank you in advance.

**Comment by TimVinkemeier** [comment]
Ah okay, of course you also have to take that limitation into account..

----- Ursprüngliche Nachricht -----
Von: "Roberto Doriguzzi Corin" notifications@github.com
Gesendet: ‎07.‎10.‎2015 14:27
An: "fp7-netide/Engine" Engine@noreply.github.com
Cc: "Tim Vinkemeier" tim@vinkemeier.de
Betreff: Re: [Engine] Testing Ryu-Backend + Core + Ryu-shim (#42)

Usually the name of the Ryu apps is defined in the init method. 
E.g.: SimpleSwitch for the OF1.0 learning switch, SimpleSwitch13 for the OF1.3 learning switch, etc.

**Comment by sergiotamu** [comment]
Hello @doriguzzi  and @TimVinkemeier ,

I have changed the module name in the CompositionSpecification to "SimpleSwitch". And now I do the same procedure but I don't use the "announcement" command. The Core shows the same error:

2015-10-06 22:08:51,833 | ERROR | pool-6-thread-7 | CompositionManager | 67 - core.caos - 1.0.0.0 | Could not handle incoming message due to configuration error.

Any idea to solve it??. Thanks =)

**Comment by TimVinkemeier** [comment]
Hi, I need more information to help, could you add a longer log-output here?

----- Ursprüngliche Nachricht -----
Von: "Sergio Tamurejo" notifications@github.com
Gesendet: ‎07.‎10.‎2015 15:08
An: "fp7-netide/Engine" Engine@noreply.github.com
Cc: "Tim Vinkemeier" tim@vinkemeier.de
Betreff: Re: [Engine] Testing Ryu-Backend + Core + Ryu-shim (#42)

Hello @doriguzzi and @TimVinkemeier ,
I have changed the module name in the CompositionSpecification to "SimpleSwitch". And now I do the same procedure but I don't use the "announcement" command. The Core shows the same error:
2015-10-06 22:08:51,833 | ERROR | pool-6-thread-7 | CompositionManager | 67 - core.caos - 1.0.0.0 | Could not handle incoming message due to configuration error.
Any idea to solve it??. Thanks =)

**Comment by sergiotamu** [comment]
Hi again @TimVinkemeier ,

Yes of course. This is my Composition Specification:
![compositionfile](https://cloud.githubusercontent.com/assets/9716567/10339031/a45894de-6d09-11e5-9027-505ab094d9cc.png)

Here you can find a longer log-output and the error I get when the Ryu shim sends the messages to the Core (two images, the second one is the continuation of the first image):
![error_1](https://cloud.githubusercontent.com/assets/9716567/10339093/ec9af994-6d09-11e5-84cb-2e11bdf42f37.png)
![error_2](https://cloud.githubusercontent.com/assets/9716567/10339100/f13b247e-6d09-11e5-88c1-2c6c85e4cb5a.png)

**Comment by TimVinkemeier** [comment]
I will have a look at them as soon as possible (hopefully this evening). Sorry, but I'm not in the office today, just answering from my phone, but this needs some more thoughts :D

----- Ursprüngliche Nachricht -----
Von: "Sergio Tamurejo" notifications@github.com
Gesendet: ‎07.‎10.‎2015 15:42
An: "fp7-netide/Engine" Engine@noreply.github.com
Cc: "Tim Vinkemeier" tim@vinkemeier.de
Betreff: Re: [Engine] Testing Ryu-Backend + Core + Ryu-shim (#42)

Hi again @TimVinkemeier ,
Yes of course. This is my Composition Specification:

Here you can find a longer log-output and the error I get when the Ryu shim sends the messages to the Core (two images, the second one is the continuation of the first image):

**Comment by sergiotamu** [comment]
Ok, thank you very much

**Comment by sergiotamu** [comment]
Hello @doriguzzi and @TimVinkemeier,

After having solved some issues with the Ryu backend and Shim I have tested the whole architecture and some errors arise. Here I am going to describe the procedure I follow and what happens in every step:

1) I execute karaf and the emulator. Then I provide the Specification file with the emulator to the Core using the "composition" command. Finally I run the Ryu backend and this is connected to the Core (apparently this works pretty well), see the images below:

The first one is the Ryu backend Console:
![backendconsole_backend-connection](https://cloud.githubusercontent.com/assets/9716567/10381552/fe93fe44-6e1b-11e5-97e6-132859f99cfd.png)

The second one illustrates the Core console:
![coreconsole_backend-connection](https://cloud.githubusercontent.com/assets/9716567/10381559/182c9d34-6e1c-11e5-92a3-7dc162ea2690.png)

2)In this step, I execute the Ryu Shim, but no notification appears in the Core console, however we can observe that something happens in the Ryu shim console. It seems that the Shim is connected to the Core (there are no errors in this step), see the following picture:
![shimconsole_shim-execution](https://cloud.githubusercontent.com/assets/9716567/10381645/ac22fdc6-6e1c-11e5-8a8a-c8e9f775e2eb.png)

3) Finally I deploy a simple topology using mininet. And here the problems begin. The dataplane sends the messages to the Ryu shim and the shim forwards the messages to the Core. The next images display the errors in the Core Console:
![coreconsole_error1](https://cloud.githubusercontent.com/assets/9716567/10381812/1f66915c-6e1e-11e5-9c02-4c8eae0779d9.png)
![coreconsole_error2](https://cloud.githubusercontent.com/assets/9716567/10381815/238118ac-6e1e-11e5-862c-5e5cf0fee3f7.png)

This third image displays the error in the Ryu backend:
![backendconsole_error](https://cloud.githubusercontent.com/assets/9716567/10381916/e97f4358-6e1e-11e5-809a-2a164740b25c.png)

The errors are produced when the Shim begins to route the messages to the Core.

What do you think that is the problem??
Thank you for the help and sorry for the long comment =)

**Comment by doriguzzi** [comment]
Regarding the backend, the cause could be that I disabled the handshake. I've just pushed a new version with the handshake enabled.
Another thing: the order should be: core, shim, backend. The backends must be started at the end.

--rob

**Comment by sergiotamu** [comment]
Hello again @doriguzzi and @TimVinkemeier ,

I have just pulled the last version of the code and I have changed in the Ryu shim and backend The "message[0]" by the "message[1]" for capturing the right message and not the empty one.
I have followed the logical order you suggested me and two different errors arise.

1) When I run mininet and Shim layer we have in the Core console this error:
![coreconsole_errors](https://cloud.githubusercontent.com/assets/9716567/10391123/1a1169f2-6e7d-11e5-8eef-66d10409743f.png)

2) After having executed the mininet and Ryu shim I run the Ryu backend. 
In the Ryu backend console we have the following error message:
![backendconsole_handshaking error](https://cloud.githubusercontent.com/assets/9716567/10391057/a291ffcc-6e7c-11e5-82a0-1917c0c8eda3.png)

And now in the Core Console we obtain this log :

![coreconsole_errorafterrunbackend](https://cloud.githubusercontent.com/assets/9716567/10391504/8530c3c0-6e7f-11e5-83a6-718e2dcf9991.png)

![coreconsole_normalbehaviour](https://cloud.githubusercontent.com/assets/9716567/10393339/72bb6122-6e8f-11e5-97d3-aa97c9f262f8.png)

Once again, the errors are produced when I run mininet and the messages from de data plane are forwarded to the Core by the Ryu shim.
I don't know why this is happening?? Any idea??.

Thank you.

**Comment by doriguzzi** [comment]
I've just updated the backend so it filters out all the messages until the handshake is completed.
The new implementation waits for the HELLO answer forever (it prints a message every 10 secs). In this way the backend can be started before the shim as well.

--rob

**Comment by TimVinkemeier** [comment]
@sergiotamu In the first screenshot you can see that the core complains about a configuration error and is still waiting for 1 module to connect (i.e. to be announced by a backend). Therefore the configuration error when messages arrive.

The errors in the second screenshot are caused by messages other than PacketIns, since they cannot be handled at the moment (we did not yet discuss how they should be handled).

The third one shows a correct output. However, you can see that it says `Waiting for request with id '173' to complete...`. This means that the core sent a request to module 173 and is waiting for a message from the backend that says that the request has finished execution. I guess this is not implemented yet by any backend, since that is a problem that I mentioned in Dublin and that we didn't really solve (you can only do that manually from the emulator at the moment, it is a preliminary solution in the core for demonstration purposes).

**Comment by sergiotamu** [comment]
Hello @TimVinkemeier  and @doriguzzi ,

1) So, in order to avoid the first error I should start the Ryu backend before than Ryu shim.
2) If the core does not support other messages than PacketIns, we would have a problem there.
3) In third case, we could do this "trick" with emulator until the backends implement that feature to solve the issue.

So taking into account these three points, it would be possible to have ready the whole architecture for next Thursday??.

Thank you in advance! =)

**Comment by doriguzzi** [comment]
1) I do not think that the order matters, however we can debug it together
2) In the shim messages are handled as follows: asynchronous messages (such as packet_in, port_status, etc.) are sent to the core with MOD_ID=0. Synchronous messages (such as features_replies, statistics, etc) are sent to the core with MOD_ID=backend_id so that the core can forward the message to the backend that generated the request (this implies that the core is also able to forward the requests with the backend_id in the header)
3) If we start with the simplest case (composition based on datapath_ids), could the core simply send the replies down to the shim without waiting?

**Comment by schwabe** [comment]
I am trying to debug this but in my test the ryu backend does not accounce any module mapping to the core

**Comment by doriguzzi** [comment]
Are you sure that the backend is using the right port? The default port set in the ryu-backend is 41414. Just look at the beginning of the code.

**Comment by sergiotamu** [comment]
Hello @schwabe 

In Ryu backend and in Ryu shim you have to replace the port by default (that Roberto uses to connect the backend and shim to the python core) by the port = 5555.

**Comment by schwabe** [comment]
I am starting the backend like this:

```
ryu-manager --ofp-tcp-listen-port 7733 ryu-backend.py tests/simple_switch.py
loading app ryu-backend.py
loading app tests/simple_switch.py
loading app ryu.controller.ofp_handler
instantiating app tests/simple_switch.py of SimpleSwitch
instantiating app ryu-backend.py of RYUClient
backend supported protocols:  {17: [1, 3, 4], 18: []}
RYU Client initiated
Connecting to Core on 127.0.0.1:5555...
instantiating app ryu.controller.ofp_handler of OFPHandler
```

Output from the core: 

```
2015-10-22 15:07:22,058 | INFO  | rint Extender: 3 | CompositionManager               | 67 - core.caos - 1.0.0.0 | Set composition finished.
2015-10-22 15:07:22,058 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | Waiting for required modules to connect, 1 left...
2015-10-22 15:07:22,059 | INFO  | rint Extender: 3 | CompositionManager               | 67 - core.caos - 1.0.0.0 | Unsupported message bypass activated.
2015-10-22 15:07:42,119 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 50 seconds remaining for 1 module(s) to connect...
2015-10-22 15:07:49,670 | INFO  | tor Receive Loop | ZeroMQBasedConnector             | 66 - core.connectivity - 1.0.0.0 | Data received from 'backend-ryu-8630'.
2015-10-22 15:07:49,675 | INFO  | tor Receive Loop | BackendManager                   | 66 - core.connectivity - 1.0.0.0 | Data received from backend 'backend-ryu-8630' with moduleId '0'.
2015-10-22 15:07:49,675 | INFO  | tor Receive Loop | BackendManager                   | 66 - core.connectivity - 1.0.0.0 | Received HelloMessage from backend, relaying to shim...
2015-10-22 15:07:49,675 | INFO  | tor Receive Loop | ZeroMQBasedConnector             | 66 - core.connectivity - 1.0.0.0 | Sending to via relay to 'shim'.
2015-10-22 15:07:49,676 | INFO  | tor Receive Loop | ZeroMQBasedConnector             | 66 - core.connectivity - 1.0.0.0 | Sent.
2015-10-22 15:08:02,162 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 40 seconds remaining for 1 module(s) to connect...
2015-10-22 15:08:22,206 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 30 seconds remaining for 1 module(s) to connect...
2015-10-22 15:08:42,251 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 20 seconds remaining for 1 module(s) to connect...
2015-10-22 15:09:02,296 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 10 seconds remaining for 1 module(s) to connect...
2015-10-22 15:09:22,337 | INFO  | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | 0 seconds remaining for 1 module(s) to connect...
2015-10-22 15:09:22,339 | ERROR | pool-11-thread-1 | CompositionManager               | 67 - core.caos - 1.0.0.0 | Reconfiguration unsuccessful: timed out with 1 unconnected modules.
java.util.concurrent.TimeoutException: Required modules did not connect.
    at eu.netide.core.caos.CompositionManager.lambda$ReconfigureAsync$30(CompositionManager.java:80)[67:core.caos:1.0.0.0]
    at eu.netide.core.caos.CompositionManager$$Lambda$3/514960542.run(Unknown Source)[67:core.caos:1.0.0.0]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_25]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_25]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_25]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_25]
    at java.lang.Thread.run(Thread.java:745)[:1.8.0_25]
```

**Comment by sergiotamu** [comment]
Hello @schwabe 

The Core seems to be waiting for the connection of a module (in this case it would be the simple_switch.py).
Firstly, have you provided a CompositionSpecification to the Core?? If you have provided to it, which composition file have you provided??

**Comment by schwabe** [comment]
Yes. I have provided this simple composition:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<CompositionSpecification xmlns="http://netide.eu/schemas/compositionspecification/v1">
  <Modules>
    <Module id="SimpleSwitch" loaderIdentification="SimpleSwitch.py"/>
  </Modules>
  <Composition>
    <ModuleCall module="SimpleSwitch" />
  </Composition>
</CompositionSpecification>
```

Simple printf debugging of ryu-backend.py also tells me that it never returns from self.handle_handshake()

**Comment by doriguzzi** [comment]
You need to run the shim and mininet as well. Otherwise the handshake cannot be completed.

**Comment by schwabe** [comment]
@doriguzzi I running all services. Is there a specific order in which the programs have to be started?

**Comment by doriguzzi** [comment]
No specific order. If the shim is not running, the backend will retry to complete the handshake every 10 secs.

**Comment by schwabe** [comment]
Okay the backend always hangs in the handle_handshake in the method message = self.socket.recv_multipart()

Shim works and the messages from the shim also reach the core: 2015-10-22 17:18:49,757 | INFO  | 

```
pool-7-thread-7  | CompositionManager               | 67 - core.caos - 1.0.0.0 | CompositionManager received message from shim: OpenFlowMessage [Header=MessageHeader [Version=VERSION_1_1,Type=OPENFLOW,Length=60,ModuleId=0,TransactionId=0,DatapathId=2],Type=PACKET_IN,OFMessage=OFPacketInVer10(xid=0, bufferId=262, totalLen=42, inPort=1, reason=NO_MATCH, data=[-1, -1, -1, -1, -1, -1, 2, -4, 102, -17, 59, -124, 8, 6, 0, 1, 8, 0, 6, 4, 0, 1, 2, -4, 102, -17, 59, -124, 10, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 0, 0, 2])]
```

**Comment by doriguzzi** [comment]
So, the backend is not receiving anything.

**Comment by schwabe** [comment]
The shim sends to be sending "random" module ids to the core which confuses my core: 

```
Sending to Core: Message header: (2, 17, 32, 0, 2803693914, 1)
Sending to Core: Message body: 04:06:00:20:a7:1c:f9:5a:00:00:00:00:00:00:00:01:00:00:01:00:fe:00:00:00:00:00:00:4f:00:00:00:00
Sending to Core: Message header: (2, 17, 32, 0, 473591711, 3)
Sending to Core: Message body: 04:06:00:20:1c:3a:6f:9f:00:00:00:00:00:00:00:03:00:00:01:00:fe:00:00:00:00:00:00:4f:00:00:00:00
Sending to Core: Message header: (2, 17, 32, 0, 844913393, 2)
Sending to Core: Message body: 04:06:00:20:32:5c:5a:f1:00:00:00:00:00:00:00:02:00:00:01:00:fe:00:00:00:00:00:00:4f:00:00:00:00
Sending to Core: Message header: (2, 17, 208, 0, 2803693916, 1)
Sending to Core: Message body: 04:13:00:d0:a7:1c:f9:5c:00:0d:00:00:00:00:00:00:ff:ff:ff:fe:00:00:00:00:c6:93:3f:81:e0:49:00:00:73:31:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:01:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:00:62:e8:ae:ab:b2:17:00:00:73:31:2d:65:74:68:31:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00:00:00:00:02:00:00:00:00:b6:5c:8d:3d:73:3b:00:00:73:31:2d:65:74:68:32:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00
Sending to Core: Message header: (2, 17, 272, 0, 473591713, 3)
Sending to Core: Message body: 04:13:01:10:1c:3a:6f:a1:00:0d:00:00:00:00:00:00:ff:ff:ff:fe:00:00:00:00:ce:63:6c:4e:e0:43:00:00:73:33:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:01:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:00:0e:a7:91:9d:cd:50:00:00:73:33:2d:65:74:68:31:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00:00:00:00:02:00:00:00:00:4a:fd:70:4a:78:e6:00:00:73:33:2d:65:74:68:32:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00:00:00:00:03:00:00:00:00:e2:d1:d1:9d:7c:29:00:00:73:33:2d:65:74:68:33:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00
Sending to Core: Message header: (2, 17, 272, 0, 844913395, 2)
Sending to Core: Message body: 04:13:01:10:32:5c:5a:f3:00:0d:00:00:00:00:00:00:ff:ff:ff:fe:00:00:00:00:1a:83:c1:05:eb:48:00:00:73:32:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:01:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01:00:00:00:00:9e:64:c8:d5:9c:d7:00:00:73:32:2d:65:74:68:31:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00:00:00:00:02:00:00:00:00:6e:e1:1f:3c:d8:32:00:00:73:32:2d:65:74:68:32:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00:00:00:00:03:00:00:00:00:7a:98:c0:2e:d0:83:00:00:73:32:2d:65:74:68:33:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:08:40:00:00:00:00:00:00:00:00:00:00:00:00:00:98:96:80:00:00:00:00
```

**Comment by doriguzzi** [comment]
The shim uses the module ids received from the core. Could you please double check that the core is sending correct values and in the big-endian byte order?

Thanks for the debug

**Comment by schwabe** [comment]
I looked in the shims code and it seems to use the xid as module id?! 

module_id = msg.xid

**Comment by ElisaRojas** [comment]
Hi all,

So I've installed tested both the Python and Java core, and the Python ones works perfectly ok, but the Java raises the same error that Sergio and Roberto came into (I'm attaching a screenshot)
![screenshot from 2015-10-29 15_50_46](https://cloud.githubusercontent.com/assets/2779496/10824417/964ef9b0-7e60-11e5-88a3-239db77e0d42.png)

I've also updated the Readme.md file from the core folder because it didn't explain the emulator part (in which you specify the composition file) and added the MinimalSpecification.xml file prepared by @sergiotamu  in the core/tools folder. If you use this file, you should execute the simple_switch.py application in the Ryu backend. (Thanks @sergiotamu for your help!)

Finally, I've captured with Wireshark both behaviours (with the Python and the Java code) and I'm going to analyse the differences now, so that we can understand why this error only occurs for the Java core. I'll let you know.

**Comment by schwabe** [comment]
@ElisaRojas what version of ryu are you running? Mine is the following:

```
# ryu-manager --version
ryu-manager 3.26
```

**Comment by ElisaRojas** [comment]
@schwabe I have 3.23 because I had it already installed from previous versions of the Engine.
@doriguzzi Do we need a specific version?
@schwabe What is your current status of the deployment? It's not very clear to me... could you tell us?

**Comment by schwabe** [comment]
My current status

shim is still sending messages with random module ids to the core, which I think is a bug in the shim (module_id = msg.xid).

backend hangs after printing:

```
instantiating app ryu.controller.ofp_handler of OFPHandler
```

**Comment by doriguzzi** [comment]
@ElisaRojas : No specific version is needed. 
@schwabe : module_id = msg.xid is inside an if clause. It is not applied to asynchronous messages such as packet_ins. @ElisaRojas: do you also see random module_ids?

@schwabe: enable the debugging messages at the beginning of the backend code, you will get more info. It is probably waiting for completing the handshake, since the Core does not communicate with the Shim

**Comment by ElisaRojas** [comment]
Hi again,

First of all @schwabe , @doriguzzi I was not paying much attention, but I think I'm not getting those random module_ids (I'm attaching a capture when using Ryu shim + Java core + Ryu backend), please let me know:
![screenshot from 2015-10-29 19_33_00](https://cloud.githubusercontent.com/assets/2779496/10828893/4695e96c-7e77-11e5-8a4a-8f96a6a5aec4.png)

Secondly, I've tried to check the wireshark traces in different scenarios:

> Note: We've been basically working with two composition files: MinimalSpecification (just a switch) + CompositionSpecification (switch and a firewall)
1. PythonCore + CompositionSpecification & JavaCore + MinimalSpecification
   - In this case I got the results shown previously (with the error at the ryu-backend). If I check the Wireshark captures, I see that after a PACKET_IN, the 5555 port sends to 39032 (the one of the backend) several TCP segments of around 70 bytes each (instead just one, or even two, containing the PACKET_IN). 
   - Conclusion: We need to analyse why this is happening (need some more time to check those traces).
2. JavaCore + CompositionSpecification
   - In this case, ryu-backend does not show anything, not even an error. If I check the captures, port 5555 never ever sends anything to any backend. I can recognise the port from the shim sending things to the core, but if I filter those in Wireshark, the is no single frame to/from 5555 to/from a different port (which should be the backend).
   - Conclusion: Maybe check how the composition specification is being managed in this example and why nothing is sent to the backend?
3. PythonCore + MinimalSpecification
   - This combination does not work... So this raises some questions for me. Maybe it's not implemented yet? I didn't capture traffic for this example.
   - Conclusion: Maybe we should focus on using the CompositionSpecification.xml file with the switch and the firewall, because we know that one works with the PythonCore at least.

Finally, currently Wireshark traces are super complex to analyse (a lot of traffic going on, probably because of ZeroMQ), so it's still a bit difficult for me to raise more conclusions than the ones I've written (with the old version of the Engine, this task was much easier :smile: ). So maybe it would be nice to have the dissector ready asap to analyse exactly what NetIDE messages are being exchanged and directly filtered by NetIDE type ( @andres-beato hint, hint :laughing: ).

That's all by now. If you need some of the Wireshark traces, please let me know.

**Comment by sergiotamu** [comment]
Hello @ElisaRojas , @doriguzzi and @schwabe 

About your first comment: In my opinion, the shim sends 'messages with random module ids to the core' when the Backend is not properly connected to the core and therefore the handshaking is incompleted. I have talked with Arne via email, and this error seems to be happening because he is executing always old compiled objects (in the Core). So, maybe the solution is to compile and run the Core as a root (with 'sudo' command).

About the second case 'JavaCore + CompositionSpecification' I think that the CompositionSpecification file is not valid, and it is impossible to assign the 'ids' to the backend and to the modules applications. It is a little bit more difficult to know what is happening in this case without more details or captures =)

**Comment by plaisthos** [comment]
@sergiotamu The old objects executed was your wrong interpretation of my comment on karaf's behaviour.

**Comment by sergiotamu** [comment]
Ok maybe I am wrong, sorry. I was just telling my opinion to solve the problem beacuse I had troubles with the Core because I was not executing the core as a root.

**Comment by schwabe** [comment]
With the fix the backend now correctly receives a packet in message but seems not to respond to it.

**Comment by ElisaRojas** [comment]
Hi again,

I updated my code (Ryu shim, backend, libraries, and both cores) and I'm bringing bad news :disappointed: But probably easy to solve:

**About Ryu+core**
Now the combination Shim+PythonCore+Backend is not working, neither with MinimalSpecification.xml nor with CompositionSpecification.xml I tried in many different ways and topologies, but still not working, sorry (the other time it worked at the first time).
@doriguzzi , I'll send you an individual email with the configuration and wireshark captures. The main idea is that now the core is not sending anything to the backend: the only messages I see from the core to the backend are the NetIDE_HELLO messages.

**About Java core**
I tried to compile and it failed:
![screenshot from 2015-11-04 21_00_50](https://cloud.githubusercontent.com/assets/2779496/10950929/ba3bdca6-833a-11e5-9403-8db4125cba3e.png)
It is probably something very easy to fix (something related to UTF-8), but I don't know how to do it, sorry. Last time it compiled perfectly the first time I did it. The error shows CompositionManager.java which is one of the files updated in the last commit.
@schwabe , can you check this? Thank you!

**Comment by doriguzzi** [comment]
Thank you Elisa. Let me finish the work I'm doing on the ryu-shim and I will try to re-produce the error. One question: is Mininet running and connected to the Ryu server?

--rob

**Comment by ElisaRojas** [comment]
@doriguzzi Yes, Mininet is running and correctly connected to the Ryu shim.
I'm about to send you the email with further clues.

**Comment by ElisaRojas** [comment]
@schwabe I've tried the new Java core code and now it compiles, but I still have the formatting problem :( Please check #51 for more info (I don't want to spam this thread), once it is solved, we can continue talking in this thread.
By the way, I haven't seen any FlowMod messages from the simple_switch.py application to the core. I'm using OVS (in Mininet) + OF1.0.

@doriguzzi Please let me know when you're done with your fixes and then I can test the ryu-python combination again.

**Comment by schwabe** [comment]
@ElisaRojas have you also reinstalled the netip java library? The bugfix was in there

**Comment by ElisaRojas** [comment]
@schwabe Yes, I've reinstalled everything (core and library), but maybe I've done it wrong. Please check #51 . I'm explaining everything there in detail and also attaching captures from Wireshark.

**Comment by ElisaRojas** [comment]
Ok, so after discussing in #51 with @schwabe , here's the summary:

The Java core generates empty 0mq messages, which make the NetIDE Wireshark dissector crash. This is not causing any problem currently in the Ryu shim/backend (they ignore them), but shouldn't be there in the future because they might cause problems with other future shims and backends, and they make it difficult debugging with Wireshark (since NetIDE can't be visualized)

With the last commit of the different modules, @schwabe sees PacketIn messages going to the backend, but I can't see them: I see PacketIns encapsulated in NetIDE from shim to core, but I don't see a similar message (same, or similar, length and content) from the core to the backend. I do not see either any FlowMod from backend to core prior to any PacketIn; I don't know if this is still happening to you @schwabe or if it was just with OF1.3 (I'm using OF1.0 and haven't tried 1.3 yet). At least NetIDE_HELLO messages are perfectly exchanged right now :)

So, @doriguzzi , please tell us when the Ryu shim/backend fixes are ready and we can continue with the testing :)

@schwabe : Add any other comment if you feel I'm missing something :)

**Comment by schwabe** [comment]
@ElisaRojas the empty messages should now be gone.

With the additional patch (pullrequest #53) to the simpleswitch to _always_ generate a flow_mod I can run the SimpleSwitch example (with a lot of warning, but it works). That is all OF1.0 at the moemnt

**Comment by ElisaRojas** [comment]
@schwabe Thank you, it's fixed now. I closed the issue.
Same for the performance issues @doriguzzi , thank you.

However, I tried to test the Ryu backend/shim with both the Java and Python core and it still not working for me. I'm attaching some screenshots from the backend and shim below. They seem to connect properly (also with Mininet), but I don't see the PacketIn messages going from the core to the backend (I see them arriving from the shim to the core though). It happens a similar thing for both cores.

![ryu-backend](https://cloud.githubusercontent.com/assets/2779496/11059786/154bc89e-879c-11e5-92a7-114eef7f60b7.png)

![ryu-shim](https://cloud.githubusercontent.com/assets/2779496/11059792/1bd9988a-879c-11e5-971f-539cd97902bb.png)

I also see some malformed NetIDE packets at the beginning #55 , but I don't think they are causing any problems at all, and I also notice the new NetIDE types, which should be implemented in the future dissector once some spec document is ready ( fp7-netide/Tools#1 ).

@doriguzzi Could you tell me if the combination is properly working for you and how is your set up and outputs from shim/backend? Just to try again :)
Thank you!

**Comment by doriguzzi** [comment]
In theory the order does not matter. Strange behaviours can happen if you re-start the core without restarting the other components.
For the log you are posting it seems that mininet is not connecting to the server controller. Could you please check that?
The problem is likely at line #49 of the netide-topo.py script: c1 = 'tcp:192.168.56.1:6633'. I must change this... In the meantime, could you please try to change it with an address that works in your environment?

**Comment by ElisaRojas** [comment]
Yes, as I said in the post, Mininet seems to be connected to the shim:
![mininet](https://cloud.githubusercontent.com/assets/2779496/11060087/36ae93c0-879e-11e5-8cb2-a46d95bede88.png)

However, I'm going to try with linear, 4 as I used to when it worked :) just in case.

**Comment by ElisaRojas** [comment]
@doriguzzi Ok, that was it, the netide-topo.py script! I tried just with linear,4 and it worked :+1: 
![mininet-linear 4](https://cloud.githubusercontent.com/assets/2779496/11060269/584706d8-879f-11e5-82d5-9a1c7c3dda4d.png)

Now I need to recheck the Wireshark traces :)

**Comment by ElisaRojas** [comment]
Ok, so I've checked the captures and this is what I'm currently seeing:
- CompositionSpecification.xml:
  The Python core exchanges between shim and backend both: FeatReq and Rep, and PacketIn and PacketOut/FlowMods. I'm also seeing SetConfig messages sent to the network from the Engine.
- MinimalSpecification.xml:
  Not working for any of the cores (neither Python, nor Java). Have you tried it @doriguzzi ?

My questions now are:
- @schwabe Is it working for you with a simple_switch app? Could you describe to me the scenario to reproduce it?
- Could I check the CompositionSpecification.xml with the Java core? Should it work? Anything special to do with the emulator?

Please leave comments regarding the current status of the Engine.
Thanks.

**Comment by schwabe** [comment]
@ElisaRojas I use the pull_request_ryu switch for the ryu components. 

Mininet (runs on a different computer):

```
sudo mn --controller=remote,ip=131.234.40.44:6633 --topo=tree,2
```

In a different terminal 

```
 sudo ovs-vsctl set bridge s1 protocols=OpenFlow10
 sudo ovs-vsctl set bridge s2 protocols=OpenFlow10
 sudo ovs-vsctl set bridge s3 protocols=OpenFlow10
```

to force OpenFlow 1.0, otherwise the ryu shim will speak  1.3

I then start the core by running bin/karaf. I do not load a compositon specification since the core remembers the last specification from its config.

Then shim:

```
ryu-manager ryu-shim.py
```

and finally 

```
ryu-manager --ofp-tcp-listen-port 7733 ryu-backend.py tests/simple_switch.py
```

after all is start, wait another 10s and then do `h1 ping h2` in the mininet console. If I let a ping running while starting the other compononents it does not work for some reason.
